google/cloud/spanner/internal/partial_result_set_reader.cc

```diff
@@ -0,0 +1,135 @@
+  }
+  if (next_value_index_ >= values_.size()) {
+    // Ran out of buffered values - try to read some more from gRPC.
+    next_value_index_ = 0;
+    auto status = ReadFromStream();
+    if (!status.ok()) {
+      return status;
+    }
+    if (finished_) {
+      return optional<Value>();
+    }
+    if (values_.empty()) {
+      return Status(StatusCode::kInternal, "response has no values");
+    }
+  }
+
+  // The metadata tells us the sequence of types for the field Values;
+  // when we reach the end of the sequence start over at the beginning.
+  auto const& fields = metadata_->row_type().fields();
+  if (next_value_type_index_ >= fields.size()) {
+    next_value_type_index_ = 0;
+  }
+
+  return {FromProto(fields.Get(next_value_type_index_++).type(),
+                    values_.Get(next_value_index_++))};
+}
+
+PartialResultSetReader::~PartialResultSetReader() {
+  if (!finished_) {
+    // The user didn't iterate over all the data; finish the stream on their
+    // behalf, but we have no way to communicate error status.
+    grpc::Status finish_status = grpc_reader_->Finish();
+    if (!finish_status.ok()) {
+      GCP_LOG(WARNING) << "Finish() failed in destructor: "
+                       << grpc_utils::MakeStatusFromRpcError(finish_status);
+    }
+  }
+}
+
+Status PartialResultSetReader::Initialize() {
+  auto status = ReadFromStream();
+  if (!status.ok()) {
+    return status;
+  }
+
+  // The first response must contain metadata.
+  if (!metadata_.has_value()) {
+    return Status(StatusCode::kInternal, "response contained no metadata");
+  }
+
+  // The metadata must contain row type information.
+  if (metadata_->row_type().fields().empty()) {
+    return Status(StatusCode::kInternal,
+                  "response metadata was missing row type information");
+  }
+  next_value_index_ = 0;
+  next_value_type_index_ = 0;
+  return Status();
+}
+
+Status PartialResultSetReader::ReadFromStream() {
+  google::spanner::v1::PartialResultSet result_set;
+  if (!grpc_reader_->Read(&result_set)) {
+    // Read() returns false for end of stream, whether we read all the data or
+    // encountered an error. Finish() tells us the status.
+    finished_ = true;
+    grpc::Status finish_status = grpc_reader_->Finish();
+    return grpc_utils::MakeStatusFromRpcError(finish_status);
+  }
+
+  if (result_set.has_metadata()) {
+    if (!metadata_.has_value()) {
+      metadata_ = std::move(*result_set.mutable_metadata());
+    } else {
+      GCP_LOG(WARNING) << "Unexpectedly received two sets of metadata";
+    }
+  }
+
+  if (result_set.has_stats()) {
+    // We should only get stats once; if not, use the last one.
+    if (stats_.has_value()) {
+      GCP_LOG(WARNING) << "Unexpectedly received two sets of stats";
+    }
+    stats_ = std::move(*result_set.mutable_stats());
+  }
+
+  // TODO(#271) store and use resume_token.
+  // TODO(#270) handle chunked_value.
+  values_ = std::move(*result_set.mutable_values());
```

Comment:
Since the type of this is `RepeatedPtrField<X>`, does it actually do an efficient move here? I tried to look it up in proto docs, and got lost. Do you know?

Reply:
My reading is that (if src and destination are on the same arena) move assignment is implemented as a swap, where the contents of a `RepeatedPtrField` are (after the arena pointer) just a size, a capacity, and a pointer to the data.